### PR TITLE
fix: http2 queueing

### LINKED ISSUE #3753: HTTP/2 GOAWAY event crashes application (Undici v6.20.1)
## Bug Description

In release 6.20.1, there appears to be a regression in http/2 support which crashes Node with a TypeError on handling HTTP/2 GoAway events. It appears to be an error introduced by [this change](https://github.com/nodejs/undici/commit/d6c44f3b49949dbdd66d9c56e729d503e6b09ddc) (specifically [here](https://github.com/nodejs/undici/blob/30305063937a87e70561d2aeb5795788ed628ca2/lib/dispatcher/client-h2.js#L201)), which adds a call to `util.errorRequest(client, request, err)` when `request` is undefined. To be honest, I'm not sure if the root of the problem is that `request` is undefined and it shouldn't be, or if that's an allowed scenario which just needs an undefined check.

The problem does not appear in 6.20.0, and I added a simple check for undefined before calling `util.errorRequest(client, request, err)` in 6.20.1 which prevents the app from crashing but I'm not familiar enough with the internals here to know if this is an acceptable solution or a deeper investigation is required to understand why `request` is undefined.

## Reproducible By

This is reproducible in my project/environment by simply waiting 6-7 minutes after issuing some http/2 requests. A 'goaway' event is eventually received and the app crashes. Unfortunately this is a large project in a private repo which I can't share. If it's not obvious to the maintainers what the problem is based on the description/screenshots here, I will be happy to try and create a simple shareable repro. 

## Expected Behavior

Receiving a goaway event on an http/2 session should not crash the application.

## Logs & Screenshots

Screenshot 2024-10-21 at 1 53 36 PM

## Environment

Node v22.10.0, Docker node:22-alpine image

### PR DESCRIPTION
**Pull request by metcoder95**
## This relates to...

Fixes #3753 

## Status

- [ ] I have read and agreed to the [Developer's Certificate of Origin][cert]
- [ ] Tested
- [ ] Benchmarked (**optional**)
- [ ] Documented
- [ ] Review ready
- [ ] In review
- [ ] Merge ready

[cert]: https://github.com/nodejs/undici/blob/main/CONTRIBUTING.md

### CODE REVIEW

**Review by ronag — APPROVED** [review]

File: test/http2.js [hunk 1]
```diff
@@ -1261,10 +1268,130 @@ test('#2364 - Concurrent aborts', async t => {
...
+
+test('#2364 - Concurrent aborts (2nd variant)', async t => {
+  const server = createSecureServer(pem)
+  let counter = 0
+
+  server.on('stream', (stream, headers, _flags, rawHeaders) => {
+    counter++
+
+    // TODO: fix this
+    // While destroyed, the stream is not handling correctly the goaway
```
Comment (metcoder95) [inline test/http2.js:1353]:
```suggestion
```

File: lib/dispatcher/client-h2.js [hunk 2]
```diff
@@ -318,14 +321,19 @@ function writeH2 (client, request) {
     util.errorRequest(client, request, err)
 
     if (stream != null) {
-      util.destroy(stream, err)
+      // On Abort, we close the stream to send RST_STREAM frame
+      stream.close()
+      // We delay the destroy to allow the stream to send the RST_STREAM frame
+      queueMicrotask(() => util.destroy(stream, err))
```
Comment (ronag) [inline lib/dispatcher/client-h2.js:327]:
```suggestion
      setImmediate(() => util.destroy(stream, err))
```

File: lib/dispatcher/client-h2.js [hunk 3]
```diff
@@ -318,14 +321,19 @@ function writeH2 (client, request) {
     util.errorRequest(client, request, err)
 
     if (stream != null) {
-      util.destroy(stream, err)
+      // On Abort, we close the stream to send RST_STREAM frame
+      stream.close()
+      // We delay the destroy to allow the stream to send the RST_STREAM frame
+      setImmediate(() => util.destroy(stream, err))
```
Comment (metcoder95) [inline lib/dispatcher/client-h2.js:327]:
```suggestion
      queueMicrotask(() => util.destroy(stream, err))
```

**Comment by metcoder95** [comment]
Can I have some support with the merge? For some reason, even if approved, I cannot merge it

**Comment by Uzlopak** [comment]
Here my approval but without having reviewed it. Just be responsible with it :P

**Pull request merged**